alibi/utils/distance.py

```diff
@@ -243,3 +242,78 @@ def multidim_scaling(d_pair: dict,
         feature_range = new_feature_range
 
     return d_abs_scaled, feature_range
+
+
+def squared_pairwise_distance(x: np.ndarray, y: np.ndarray, a_min: float = 1e-30, a_max: float = 1e30) -> np.ndarray:
+    """
+    `numpy` pairwise squared Euclidean distance between samples `x` and `y`.
+
+    Parameters
+    ----------
+    x
+        A batch of instances of shape `Nx x features`.
+    y
+        A batch of instances of shape `Ny x features`.
+    a_min
+        Lower bound to clip distance values.
+    a_max
+        Upper bound to clip distance values.
+
+    Returns
+    -------
+    Pairwise squared Euclidean distance `Nx x Ny`.
+    """
+    x2 = np.sum(x ** 2, axis=-1, keepdims=True)
+    y2 = np.sum(y ** 2, axis=-1, keepdims=True)
+    dist = x2 + np.transpose(y2, (1, 0)) - 2. * x @ np.transpose(y, (1, 0))
+    return np.clip(dist, a_min=a_min, a_max=a_max)
+
+
+def batch_compute_kernel_matrix(x: np.ndarray,
+                                y: np.ndarray,
+                                kernel: Callable[[np.ndarray, np.ndarray], np.ndarray],
+                                batch_size: int = int(1e10),
+                                preprocess_fn: Optional[Callable[[np.ndarray], np.ndarray]] = None) -> np.ndarray:
+    """
+    Compute the kernel matrix between `x` and `y` by filling in blocks of size
+    `batch_size x batch_size` at a time.
+
+    Parameters
+    ----------
+    x
+        The first array of data instances.
+    y
+        The second array of data instances.
+    kernel
+        Kernel function to be used for kernel matrix computation.
+    batch_size
+        Batch size to be used for each prediction.
+    preprocess_fn
+        Optional preprocessing function for each batch.
+
+    Returns
+    -------
+    Kernel matrix in the form of a `numpy` array.
+    """
+    if type(x) != type(y):
+        raise ValueError("x and y should be of the same type")
```

Comment:
Do we expect this to realistically arise in practice given that this is a function primarily for internal use and that the type hints would only accept `x` and `y` to be `np.ndarray`?

Reply:
also inherited from [detect](https://github.com/SeldonIO/alibi-detect/blob/a0e7082ebb12206fa7f5417cab0353c55169cfa3/alibi_detect/utils/tensorflow/distance.py#L62-L63)